bip-alfredhodler-privatepayments.mediawiki

```diff
@@ -0,0 +1,251 @@
+
+==Specification==
+
+===Definitions===
+
+* Alice: sender
+* Bob: recipient
+* Payment code: static string that Bob generates and shares with others so that he can receive payments
+* ''P'': public key contained in Bob's payment code
+* ''p'': private key associated with Bob's public key ''P''
+* ''N'': extended public key used by Alice to derive child keys for each Bob she wants to transact with
+* ''n'': private key associated with Alice's public key ''N''
+* ''x'': Alice's secret recipient index, unique for each Bob
+* ''N<sub>x</sub>'': child public key derived from ''N'' at index ''x'' (non-hardened)
+* ''n<sub>x</sub>'': private key associated with ''N<sub>x</sub>''
+* ''c'': Alice's transaction count toward Bob
+* ''P<sub>c</sub>'': Bob's public key at index ''c''
+* ''p<sub>c</sub>'': Bob's private key at index ''c''
+* ''A<sub>c</sub>'': Bob's receive address at index ''c''
+* ''H'': SHA256 hash function
+* ''*'': EC multiplication
+* ''+'': EC addition
+* ''|'': string concatenation
+* ''[a..b]'': string slicing (inclusive of ''a'', exclusive of ''b'')
+
+===Public Key Derivation Path===
+
+The derivation path for this BIP follows BIP44. The following BIP32 path levels are defined:
+
+<code>
+m / purpose' / coin_type' / account'
+</code>
+
+<code>purpose</code> is set to 999.
+
+''(p, P)'' and ''(n, N)'' are keys associated with the above path, depending on which side is performing the calculation.
+
+''N<sub>x</sub>'' keys are the direct non-hardened children of ''N''. For instance, the path of ''N<sub>0</sub>'' from ''N'' is ''m / 0''.
+
+===Payment Code Structure and Encoding===
+
+* bytes <code>[0..2]</code>: address type flags (2 bytes)
+* bytes <code>[2..36]</code>: compressed public key P (33 bytes)
+
+Payment codes are encoded in bech32m and the human readable part is "pay" for mainnet and "payt" for testnet (all types), resulting in payment codes that look like "pay1cqqq8d29g0a7m8ghmycqk5yv24mfh3xg8ptzqcn8xz6d2tjl8ccdnfkpjl7p84".
+
+===Address Types===
+
+Address type flags determine which address types a payment code accepts. This is represented by big-endian ordered 16 bits. For instance, a hypothetical payment code that handles all address types will have all defined bits set to 1 (<code>0xffff</code>).
+
+Currently defined flags:
+
+{| class="wikitable"
+! Address Type !! Flag !! Flag Value !! Ordinal Value
+|-
+| P2PKH || <code>1 << 0</code> || <code>0x0001</code> || 0
+|-
+| P2WPKH || <code>1 << 1</code> || <code>0x0002</code> || 1
+|-
+| P2TR || <code>1 << 2</code> || <code>0x0004</code> || 2
+|}
+
+The remaining flags are reserved for future address types.
+
+While payment codes use 2-byte bitflag arrays, notifications use ordinal values in the form of a single byte.
+
+===Notifications===
+
+Notifications are performed by publishing transactions that contain a 40-byte <code>OP_RETURN</code> output. The value of the <code>OP_RETURN</code> is constructed using the following formula:
```

Comment:
I thought part of the point of this BIP was to avoid such spam?

Why can't this stay out of band?

Reply:
Notifications can indeed be performed out of band (the original BIP was trying to use Bitmessage), but then we lose the ability to deterministically restore a wallet from seed simply by rescanning the blockchain.

So if we were to get rid of OP_RETURNs, we'd either need a centralized directory of notifications that's vulnerable to takedowns, or each user would have to back up their set of notifications (losing it would mean losing funds forever).

I tried my best to cut the size of these notifications from 80 bytes in BIP47 to half that amount.

Reply:
This older draft didn't require notifications at all?

https://gist.github.com/RubenSomsen/c43b79517e7cb701ebf77eec6dbb46b8

Reply:
Silent Payments is a different standard by different people that I had no part in making. It requires a full node and doesn't work with light clients such as mobile wallets.

Private Payments requires these notifications in order to be compatible with light wallets.

Reply:
I'm showing `OP_RETURN` at 2.8 GB right now, which is 0.6% of the total chain size and is prunable if full nodes need that space.

The history of total `OP_RETURN` size looks like a very large user came online at one point but then stopped after a while.

![nulldata-totals-graph](https://user-images.githubusercontent.com/1562417/185493625-6887bc74-2283-4882-b9e7-6763e6c45552.png)

This protocol does not need to continually place anchors into the chain like other protocols, so the usage should have much less impact.

Reply:
Bitcoin's blockchain is not for storing private backups, or distributing information between users.

Instead of lazily abusing OP_RETURN, it would be preferable to have an external central directory, but there's no reason the directory couldn't be p2p either. It doesn't require any of the properties of the blockchain (double spend protection).

Reply:
I expanded on out of band notifications.

Reply:
Far better to just require a full node, than to spam Bitcoin for the benefit of light wallets (ie, non-Bitcoin users).

(FWIW, I would discourage implementation of this BIP it it keeps OP_RETURN abuse.)